text/0022-lazy-electron-download.md

```diff
@@ -0,0 +1,180 @@
+```
+
+## Reference-level explanation
+
+### Implementation Details
+
+1. **Remove postinstall script** from `package.json`:
+   - Remove the `postinstall` script entry
+   - Keep existing download utilities for reuse
+
+2. **Modify CLI wrapper** (`cli.js` or equivalent):
+   - Check if Electron binary exists at expected path
+   - If not present, trigger download before spawning Electron process
+   - Show progress indicator during download
+   - Handle download errors gracefully with clear error messages
+
+3. **Add `install-electron` binary**:
+   - Create new executable that explicitly triggers download
+   - Export it in `package.json` `bin` field
+   - Reuse existing download logic from current `install.js`
+
+4. **Path resolution**:
+   - Keep current binary path resolution logic
+   - Ensure download location matches existing conventions
+   - Respect `ELECTRON_CUSTOM_DIR` and other environment variables
+
+5. **Error handling**:
+   - Clear error messages if download fails
+   - Guidance on manual download if automatic download fails
+   - Network error retry logic
+
+### Code flow
+
+```javascript
+// Simplified example of CLI wrapper logic
+async function main() {
+  const { getElectronPath } = require('./path');
+  const { download } = require('./download');
+
+  let electronPath = getElectronPath();
+
+  if (!fs.existsSync(electronPath)) {
+    console.log('Electron binary not found, downloading...');
+    await download();
+    electronPath = getElectronPath();
+  }
+
+  // Spawn Electron process
+  spawn(electronPath, process.argv.slice(2), { stdio: 'inherit' });
+}
+```
+
+### Migration Path
+
+This is a breaking change only for specific edge cases. The migration path is:
+
+1. **Most users**: No action required - download happens on first run instead of install
+2. **Users with custom build scripts expecting binary at install time**: Add `install-electron` to their build process
+3. **Docker/CI environments**: Either accept lazy download or explicitly run `install-electron`
+
+## Drawbacks
+
+1. **First run delay**: Users will experience a download delay on first run instead of during installation. However, this is only a UX change, not a functional regression.
+
+2. **Breaking change for some build pipelines**: Build systems that expect the binary to exist immediately after `npm install` will need to be updated to run `install-electron` explicitly.
+
+3. **CI cache implications**: Some CI systems cache based on `npm install`. They may need to adjust caching strategies to include the lazy download step.
+
+4. **Error visibility**: Download errors move from install-time to run-time, which could be surprising for some users (though error messages can guide them and it's probably better than electron not downloading doesn't break package installs).
+
+## Alternative designs considered
+
+1. **Keep postinstall but make it optional**: Would require environment variable configuration, less clean than lazy download
```

Comment:
I'm in favor of this, developers use `pnpm@10+` will default skip running postinstall, and they are aware of this if they choose to upgrade pnpm, and for older developers we can keep the behavior to prevent break workflows. The lazy download could be a additional check if developers doesn't install the binary for whatever reason.

Reply:
The reason I excluded this was consistency, the first question in the bug report shouldn't be "ok so do you have postinstall enabled? oh you don't know? ok are you using pnpm? or yarn > 4.10?". That just doesn't found fun. I think the experience should be consistent across all package managers

Reply:
To me, being able to run `npx install-electron` on `postinstall` sounds like a happy workaround for devs who want to still run it that way